src/System.Private.CoreLib/shared/System/String.Comparison.cs

```diff
@@ -753,54 +753,64 @@ public override int GetHashCode()
         // that string.Equals(A, B, C), then they will return the same hash code with this comparison C.
         public int GetHashCode(StringComparison comparisonType) => StringComparer.FromComparison(comparisonType).GetHashCode(this);
 
-        // Use this if and only if you need the hashcode to not change across app domains (e.g. you have an app domain agile
-        // hash table).
-        internal int GetLegacyNonRandomizedHashCode()
+        // Use this if and only if 'Denial of Service' attacks are not a concern (i.e. never used for free-form user input),
+        // or are otherwise mitigated
+        internal unsafe int GetNonRandomizedHashCode()
         {
-            unsafe
+            fixed (char* src = &_firstChar)
             {
-                fixed (char* src = &_firstChar)
-                {
-                    Debug.Assert(src[this.Length] == '\0', "src[this.Length] == '\\0'");
-                    Debug.Assert(((int)src) % 4 == 0, "Managed string should start at 4 bytes boundary");
+                Debug.Assert(src[this.Length] == '\0', "src[this.Length] == '\\0'");
+                Debug.Assert(((int)src) % 4 == 0, "Managed string should start at 4 bytes boundary");
+
+                uint hash1 = (5381 << 16) + 5381;
+                uint hash2 = hash1;
+
+                int length = this.Length;
+                uint* current = (uint*)src;
+
+                char* end = (src + length);
 #if BIT64
-                    int hash1 = 5381;
-#else // !BIT64 (32)
-                    int hash1 = (5381<<16) + 5381;
+                if (length >= 4 && (((int)current) & 7) != 0)
```

Comment:
`(((int)current) & 7) != 0` is always going to be true given the current String layout on 64-bit.

Reply:
Maybe should be a Debug.Assert instead?

Reply:
I'm not sure that's true? It seems interned strings sizes 0 - 3 go down the other path?

Reply:
duh; but I test that the length is `>= 4` first...

Reply:
oh, that's probably why its going down the other path; ok, bit embarrassed...